src/test/openads-appnexus/repository/AdListenerRepositoryTest.js

```diff
@@ -0,0 +1,92 @@
+import {expect} from 'chai'
+
+import ReplayEventBus from '../../../openads-appnexus/service/ReplayEventBus'
+import AdListenerRepository from '../../../openads-appnexus/repository/AdListenerRepository'
+
+describe('AdListenerRepository', function() {
+  beforeEach(() => {
+    ReplayEventBus.clear()
+  })
+  describe('remove method', function() {
+    it('should clear the message bus for the event name related to the received id', done => {
+      const givenId = 'ad1'
+      ReplayEventBus.clear()
+      ReplayEventBus.register({
+        eventName: 'AD_STORED_ad1',
+        observer: () => null
+      })
+
+      const repository = new AdListenerRepository()
+      repository
+        .remove({id: givenId})
+        .then(() => {
+          expect(
+            ReplayEventBus.getNumberOfRegisteredEvents(),
+            'the observer has not been removed'
+          ).to.equal(0)
+          done()
+        })
+        .catch(e => done(e))
+    })
+  })
+  describe('find method', function() {
+    it('should end with a timeout exception if no ad is found within the time limit', done => {
+      const givenTimeout = 5
+      const givenId = 'notfound'
+      const repository = new AdListenerRepository({timeout: givenTimeout})
+      repository
+        .find({id: givenId})
+        .then(() => done(new Error('should not find any ad')))
+        .catch(e => {
+          expect(e.message, 'should be a timeout error').to.include(
+            'Timeout waiting'
+          )
+          done()
+        })
+        .catch(e => done(e))
+    })
+    it('should find the Ad that is raised within the AD_STORED_x event', done => {
```

Comment:
Add a test case for calling find method twice ;)
Expected result is the same value and the number of observers registered must be 0 for this event type right?

Reply:
Well, not exactly 'the same', as it does not store the result, calling find method twice over the same 'id' would result in 2 promises waiting for an event to be raised, so if the 2 find observers are registered before the Ad retrieval is raised (done when the apntag AD_AVAILABLE is raised btw) the 2 promises will be resolved at a time in the registration order (it's the expected)
If there's a find->retrieval->find, the second find will wait for the Ad to be retieved once again by the AD_AVAILABLE event